ta/arch/arm/link.mk

```diff
@@ -5,6 +5,8 @@ link-script-dep$(sm) = $(link-out-dir$(sm))/.ta.ld.d
 SIGN_ENC ?= $(ta-dev-kit-dir$(sm))/scripts/sign_encrypt.py
 TA_SIGN_KEY ?= $(ta-dev-kit-dir$(sm))/keys/default_ta.pem
 
+user-ta-version ?= 0
```

Comment:
What if the previous in-tree TA has set `user-ta-version` to some value and the current in-tree TA doesn't set it? The "wrong" value would be used, right? I think this line is OK provided that the variable is cleared at the end of this makefile, like so:

```diff
--- a/ta/arch/arm/link.mk
+++ b/ta/arch/arm/link.mk
@@ -103,3 +103,4 @@ endef
 $(eval $(call gen-link-t))

 additional-link-deps :=
+user-ta-version :=
```

Then, I'd like the commit message to describe more in details what's happening (the subject is OK though). My suggestion:

```
When a user TA is built with the TA dev kit (out-of-tree TA),
user-ta-version is always set to the value of CFG_TA_VERSION (default:
0) by ta_dev_kit.mk. But for in-tree TAs, this variable is never set
outside the in-tree TA makefiles (ta/*/user_ta.mk). Therefore it is
mandatory for in-tree uer TAs to set the variable or various issues
will happen (build error or unexpected version taken from the TA that
happened to have been built just before).

To address the problem and for consistency this commit sets a default
value of 0 for user-ta-version in ta/arch/arm/link.mk and makes sure
the variable is cleared each time a new user TA gets built.

Fixes: ...
Signed-off-by: ...
```

Reply:
Clearing user-ta-version that way does not work, it breaks the build again.

Reply:
Maybe another way could be to add argument `--ta-version $(user-ta-version)` only upon `user-ta-version` definition. Something like.

```patch
 ifeq ($(CFG_ENCRYPT_TA),y)
 crypt-args$(user-ta-uuid) := --enc-key $(TA_ENC_KEY)
 cmd-echo$(user-ta-uuid) := SIGNENC
 endif
+ifdef user-ta-version
+user-ta-version-args$(user-ta-uuid) := --ta-version $(user-ta-version)
+endif
 $(link-out-dir$(sm))/$(user-ta-uuid).ta: \
 			$(link-out-dir$(sm))/$(user-ta-uuid).stripped.elf \
 			$(TA_SIGN_KEY)
 	@$(cmd-echo-silent) '  $$(cmd-echo$(user-ta-uuid)) $$@'
 	$(q)$(SIGN_ENC) --key $(TA_SIGN_KEY) $$(crypt-args$(user-ta-uuid)) \
-		--uuid $(user-ta-uuid) --ta-version $(user-ta-version) \
+		--uuid $(user-ta-uuid) $$(user-ta-version-args$(user-ta-uuid)) \
 		--in $$< --out $$@
```

Reply:
Found the solution, update will follow.

Reply:
OK I'm stopping the guesswork here :wink: